crates/core/src/factories/core_fetch.rs

```diff
@@ -0,0 +1,328 @@
+                    self.0.fetch_queue_tx.try_send((op_id, source.clone()))
+                {
+                    tracing::warn!(
+                        "could not pass fetch request to fetch task: {err}"
+                    );
+                }
+            });
+
+            Ok(())
+        })
+    }
+}
+
+type FetchRequest = (OpId, AgentId);
+
+#[derive(Debug)]
+struct State {
+    ops: HashSet<FetchRequest>,
+    cool_down_list: CoolDownList,
+}
+
+#[derive(Debug)]
+struct Inner {
+    state: Arc<Mutex<State>>,
+    fetch_queue_tx: Sender<FetchRequest>,
+    fetch_queue_rx: Arc<Mutex<Receiver<FetchRequest>>>,
+}
+
+impl Inner {
+    pub fn new(cool_down_interval_ms: u64) -> Self {
+        // Create a channel to send new ops to fetch to the tasks. This is in effect the fetch queue.
+        let (fetch_queue_tx, fetch_queue_rx) = channel::<FetchRequest>(1024);
+        let fetch_queue_rx = Arc::new(Mutex::new(fetch_queue_rx));
+
+        let ops = HashSet::new();
+        let cool_down_list = CoolDownList::new(cool_down_interval_ms);
+
+        Self {
+            state: Arc::new(Mutex::new(State {
+                ops,
+                cool_down_list,
+            })),
+            fetch_queue_tx,
+            fetch_queue_rx,
+        }
+    }
+
+    pub fn spawn_fetch_tasks(
+        &self,
+        parallel_request_count: u8,
+        transport: DynTransport,
+    ) {
+        for _ in 0..parallel_request_count {
+            tokio::task::spawn(Inner::fetch_task(
+                self.fetch_queue_tx.clone(),
+                self.fetch_queue_rx.clone(),
+                transport.clone(),
+                self.state.clone(),
+            ));
+        }
+    }
+
+    async fn fetch_task(
+        fetch_request_tx: Sender<FetchRequest>,
+        fetch_request_rx: Arc<Mutex<Receiver<FetchRequest>>>,
+        transport: DynTransport,
+        state: Arc<Mutex<State>>,
+    ) {
+        while let Some((op_id, agent_id)) =
+            fetch_request_rx.lock().await.recv().await
+        {
+            // Ensure op is still in the set of ops to fetch.
+            if !state
+                .lock()
+                .await
+                .ops
+                .contains(&(op_id.clone(), agent_id.clone()))
+            {
+                continue;
+            }
+
+            // Check if agent is on cool-down list.
+            if !state
+                .lock()
+                .await
+                .cool_down_list
+                .is_agent_cooling_down(&agent_id)
+            {
+                // Send fetch request to agent.
+                // TODO: update when transport is available
+                let mut data = bytes::BytesMut::new();
+                data.put(op_id.clone().0 .0);
+                data.put(agent_id.clone().0 .0);
+                let data = data.freeze();
+                if let Err(err) = transport
+                    .send_module(
+                        Url::from_str("wss://0.0.0.0:443").unwrap(),
```

Comment:
Heh... woo, there's a lot to figure out here : ) -- maybe doesn't have to be part of this PR, but we'll need to query the peer_store for a url to set here.

Reply:
I added the peer store to the factory create method, but got stuck in the tests. https://github.com/holochain/kitsune2/pull/35/commits/f90981e0b69f31213ec7f27e682ec7fd1e889d06